Do not rely on length of sys.argv

samdoran opened the pull request:
From the current behavior, it seems the goal is to print a common usage message and exit 0 if no sub command was given. Under other circumstances, such as an invalid or missing parameter, exit with 2 and print the normal usage message.

In order to preserve this behavior and make the tests run consistently whether or not they are run in multiple process, a custom error method was created that inspects the error message and exits 0 with the common usage message if no command was given. This is a bit fragile, but inside that method we do not have easy access to arguments passed to the main() function (they could be inferred by inspecting private attributes, but this is probably unwise).

This requires catching the SystemExit in the main() function and deciding there based on sys_args how to handle it.

This fix preservers the existing behavior whil making the tests pass consistently. However, it would be much simpler to rely on the default behavior of argparse and exit 2 in both cases. This would remove the need for a custom error method and exception handling. It would be a change in behavior, though.

Related to #815.

samdoran committed:
Do not rely on length of sys.argv

From the current behavior, it seems the goal is to print a common usage message
and exit 0 if no sub command was given. Under other circumstances, such as an
invalid or missing parameter, exit with 2 and print the normal usage message.

In order to preserve this behavior and make the tests run consistently whether
or not they are run in multiple process, a custom error method was created
that inspects the error message and exits 0 with the common usage message
if no command was given. This is a bit fragile, but inside that method
we do not have easy access to arguments passed to the main() function (they
could be inferred by inspecting private attributes, but this is probably
unwise).

This requires catching the SystemExit in the main() function and deciding there
based on sys_args how to handle it.

This fix preservers the existing behavior whil making the tests pass
consistently. However, it would be much simpler to rely on the default behavior
of argparse and exit 2 in both cases. This would remove the need for a custom
error method and exception handling. It would be a change in behavior, though.

samdoran commented:
I added tests to ensure 100% coverage of the new code.

AlanCoding commented on ansible_runner/__main__.py:
```diff
@@ -754,12 +765,15 @@ def main(sys_args=None):
 
-    args = parser.parse_args(sys_args)
+        # Otherwise, assume a call to main() passed in invalid arguments
+        parser.exit(2)
```
wouldn't `raise` work in this case too? Invalid arguments are already processed correctly as it stands, which works as a `SystemExit` exception uncaught here from `parser.parse_args(sys_args)`

samdoran replied:
Yeah, this whole thing is a bit convoluted. I don't love it.

The reason `raise` won't work is that the custom `error()` method changes the exit code. And, at least my from understanding of the existing code and tests, it should raise 2 when no args are passed in the Python interface but 0 when no args are passed in the CLI.

I'm really hoping someone will tell me "Oops we didn't mean to do that. It's fine to just exit with 2 in all cases."

```
> ansible-runner > /dev/null 2>&1 ; echo $status
0
> python -c 'from ansible_runner.__main__ import main; main([])' > /dev/null 2>&1 ; echo $status
2
```

If I change it to raise, the Python interface behavior changes.

```
> python -c 'from ansible_runner.__main__ import main; main([])' > /dev/null 2>&1 ; echo $status
0
```

AlanCoding replied:
The 0 exit code looks like it was introduced in #333

I offer no defense, whatsoever, to giving an exit code of 0 for `ansible-runner` by itself. A code of 2 sounds _more_ correct. I care about that less than I care about making this code clean and maintainable.

AlanCoding replied:
@matburt do you have any reason to keep the exit codes as they are?

samdoran replied:
I got same feedback from @shanemcd and it seems we're ok with exiting 2 in all cases. I will work on simplifying this PR in a bit.

samdoran replied:
The latest changes are _much_ cleaner.

samdoran committed:
Always exit 2 for incorrect args

Only print out common help with no sub command was given

samdoran committed:
Monkeypatch sys.argv in the test

When running with -n0, the arguments passed to pytest leak into the test,
causing different behavior.

Shrews commented:
recheck
